toqito/state_opt/npa_hierarchy.py

```diff
@@ -6,119 +6,225 @@
 import cvxpy
 
 Symbol = namedtuple("Symbol", ["player", "question", "answer"], defaults=["", None, None])
+IDENTITY_SYMBOL = Symbol("", None, None)  # Explicit identity symbol
 
 
-# This function simplifies the input word by applying
-# the commutation and projector rules.
-def _reduce(word: tuple[Symbol]) -> tuple[Symbol]:
-    # commute: bring Alice in front.
-    w_a, w_b = (), ()
-    for symbol in word:
-        if symbol.player == "Alice":
-            w_a += (symbol,)
-        if symbol.player == "Bob":
-            w_b += (symbol,)
+def _reduce(word: tuple[Symbol, ...]) -> tuple[Symbol, ...]:
+    """Reduce an operator word to its canonical form using NPA rules.
 
-    word = w_a + w_b
-    for i in range(len(word) - 1):
-        symbol_x, symbol_y = word[i], word[i + 1]
-
-        # projector: merge them.
-        if symbol_x == symbol_y:
-            return _reduce(word[:i] + word[i + 1 :])
+    Identity: I*S = S*I = S, I*I = I
+    Commutation: Alice operators commute with Bob operators. Canonical form: A...AB...B
+    Orthogonality: P_x,a P_x,b = 0 if a != b (for same player x)
+    Idempotence: P_x,a P_x,a = P_x,a (for same player x)
+    """
+    if not word:
+        return ()
+
+    # Initial pass to filter out identities IF other ops are present
+    current_list = [s for s in word if s != IDENTITY_SYMBOL]
+    if not current_list:  # Original word was all identities or empty
+        return (IDENTITY_SYMBOL,) if any(s == IDENTITY_SYMBOL for s in word) else ()
+
+    # Canonical player order (Alice then Bob), preserving original relative internal order
+    alice_ops = [s for s in current_list if s.player == "Alice"]
+    bob_ops = [s for s in current_list if s.player == "Bob"]
+    current_list = alice_ops + bob_ops  # This is now a list of Symbol objects
+
+    # Iteratively apply reduction rules until no more changes occur
+    while True:
+        len_before_pass = len(current_list)
+        next_pass_list = []
+        idx = 0
+        made_change_in_pass = False
+
+        while idx < len(current_list):
+            s_x = current_list[idx]
+
+            if idx + 1 < len(current_list):
+                s_y = current_list[idx + 1]
+                # Only apply if s_x and s_y are from the same player.
+                if s_x == s_y and s_x.player in ("Alice", "Bob"):  # s_x != IDENTITY_SYMBOL
```

Comment:
We seem to check for whether something is a valid player symbol in this way in a few different areas of our code. That is, we seem to have some repeated instances of `X in ("Alice", "Bob")`. In the interest in reducing repeated code and also in the event where we may consider games with more players, what do you think about defining something similar to what you do for `IDENTITY_SYMBOL`, but this time, for the players?

This is just a suggestion, so feel free to disagree and not implement if you wish!